File: java/src/main/java/it/giannibombelli/wsc2026/booking/domain/booking/Booking.java
package it.giannibombelli.wsc2026.booking.domain.booking;

import it.giannibombelli.wsc2026.common.utils.Require;

import it.giannibombelli.wsc2026.booking.domain.events.BookingResultEvents;
import it.giannibombelli.wsc2026.booking.domain.primitive.GiftCardReference;
import it.giannibombelli.wsc2026.common.domain.model.Aggregate;
import it.giannibombelli.wsc2026.common.domain.primitive.Description;
import it.giannibombelli.wsc2026.common.domain.primitive.Money;

import java.util.Objects;

import static it.giannibombelli.wsc2026.booking.domain.events.BookingResultEvents.*;


public final class Booking implements Aggregate<BookingId> {
    private final BookingId id;
    private final Description description;
    private final GiftCardReference giftCardReference;
    private BookingStatus status;

    public Booking(BookingId id, Description description, GiftCardReference giftCardReference, BookingStatus status) {
        Require.requireArgument(id, "id");
        Require.requireArgument(description, "description");
        Require.requireArgument(giftCardReference, "giftCardReference");
        Require.requireArgument(status, "status");

        this.id = id;
        this.description = description;
        this.giftCardReference = giftCardReference;
        this.status = status;
    }

    public static Booking place(BookingId id, Description description, GiftCardReference giftCardReference) {
        Require.requireArgument(id, "id");
        Require.requireArgument(description, "description");
        Require.requireArgument(giftCardReference, "giftCardReference");

        return new Booking(id, description, giftCardReference, BookingStatus.PLACED);
    }

    public BookingResultEvents confirm(Money amount) {
        Require.requireArgument(amount, "amount");

        // TODO: implement business rule for confirmation vs refusal
        // (e.g. check room availability, overbooking policy, etc.)
        boolean canConfirm = true;

        if (canConfirm) {
            status = BookingStatus.CONFIRMED;
            return new BookingConfirmed(this.id, this.giftCardReference.value(), amount);
        } else {
            status = BookingStatus.REFUSED;
            return new BookingRefused(this.id, this.giftCardReference.value(), amount);
        }
    }

    public BookingRejected reject(Money amount) {
        Require.requireArgument(amount, "amount");

        status = BookingStatus.REJECTED;
        return new BookingRejected(this.id, this.giftCardReference.value(), amount);
    }

    public BookingId id() {
        return id;
    }

    public Description description() {
        return description;
    }

    public GiftCardReference giftCardReference() {
        return giftCardReference;
    }

    public BookingStatus status() {
        return status;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        return Objects.equals(id, ((Booking) o).id);
    }

    @Override
    public int hashCode() {
        return Objects.hashCode(id);
    }

    @Override
    public String toString() {
        return "Booking[" +
            "id=" + id + ", " +
            "description=" + description + ", " +
            "giftCardReference=" + giftCardReference + ", " +
            "status=" + status +
            ']';
    }
}

File: typescript/src/booking/domain/booking/booking.ts
// Aggregato Booking: gestisce il ciclo di vita di una prenotazione.

// Refactor to use a published language (e.g., common types or primitive IDs).

import type { Aggregate } from '@/common/domain/model/aggregate.js';
import { Description } from '@/common/domain/primitive/description.js';
import { Money } from '@/common/domain/primitive/money.js';
import { GiftCardReference } from '../primitive/giftCardReference.js';
import type { BookingResultEvent, BookingRejected } from '../events/bookingResultEvents.js';
import { bookingConfirmed, bookingRefused, bookingRejected } from '../events/bookingResultEvents.js';
import { BookingId } from './bookingId.js';
import { BookingStatus } from './bookingStatus.js';
import { requireArgument } from '@/common/utils/requireArgument.js';

export class Booking implements Aggregate<BookingId> {
  private readonly _id: BookingId;
  private readonly _description: Description;
  private readonly _giftCardReference: GiftCardReference;
  private _status: BookingStatus;

  constructor(id: BookingId, description: Description, giftCardReference: GiftCardReference, status: BookingStatus) {
    requireArgument(id, 'id');
    requireArgument(description, 'description');
    requireArgument(giftCardReference, 'giftCardReference');
    requireArgument(status, 'status');

    this._id = id;
    this._description = description;
    this._giftCardReference = giftCardReference;
    this._status = status;
  }

  static place(id: BookingId, description: Description, giftCardReference: GiftCardReference): Booking {
    requireArgument(id, 'id');
    requireArgument(description, 'description');
    requireArgument(giftCardReference, 'giftCardReference');

    return new Booking(id, description, giftCardReference, BookingStatus.PLACED);
  }

  confirm(amount: Money): BookingResultEvent {
    requireArgument(amount, 'amount');

    // TODO: implement business rule for confirmation vs refusal
    // (e.g. check room availability, overbooking policy, etc.)
    const canConfirm = true;

    if (canConfirm) {
      this._status = BookingStatus.CONFIRMED;
      return bookingConfirmed(this._id, this._giftCardReference.value.value, amount);
    } else {
      this._status = BookingStatus.REFUSED;
      return bookingRefused(this._id, this._giftCardReference.value.value, amount);
    }
  }

  reject(amount: Money): BookingRejected {
    requireArgument(amount, 'amount');

    this._status = BookingStatus.REJECTED;
    return bookingRejected(this._id, this._giftCardReference.value.value, amount);
  }

  id(): BookingId {
    return this._id;
  }

  description(): Description {
    return this._description;
  }

  giftCardReference(): GiftCardReference {
    return this._giftCardReference;
  }

  status(): BookingStatus {
    return this._status;
  }
}
